File: equal_row_col.rb
# @param {Integer[][]} grid
# @return {Integer}
def equal_pairs(grid)
    row_hash = Hash.new { |hash, key| hash[key] = [] }
    count = 0
  
    grid.each_with_index do |row, i|
      row_hash[row] << i
    end
  
    grid[0].size.times do |j|
      col = (0...grid.size).map { |k| grid[k][j] }
      count += row_hash[col].size
    end
  
    count
end
  
# bruteforce soln
# @param {Integer[][]} grid
# @return {Integer}

def equal_pairs(grid)
    count = 0
    (0...grid.size).each do |i|
        (0...grid[0].size).each do |j|
          count += 1 if grid[i] == (0...grid[0].size).map { |k| grid[k][j] }
        end
    end
    count
end

File: equal_row_col.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
const arrayEquals = (arr1, arr2) => {
  return arr1.every((el, idx) => el === arr2[idx]);
};

var equalPairs = function (grid) {
  let result = 0;
  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[i].length; j++) {
      const row = grid[i];
      const col = [];

      for (let k = 0; k < grid.length; k++) {
        col.push(grid[k][j]);
      }
      if (arrayEquals(row, col)) {
        result++;
      }
    }
  }

  return result;
};
